sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GoneAndRetryWithRetryPolicy.java

```diff
@@ -16,189 +15,324 @@
+
+        private final RxDocumentServiceRequest request;
+        private volatile int attemptCount = 1;
+        private volatile int attemptCountInvalidPartition = 1;
+        private volatile int currentBackoffSeconds = GoneRetryPolicy.INITIAL_BACKOFF_TIME;
+        private final StopWatch durationTimer;
+        private final int waitTimeInSeconds;
+        //TODO once this is moved to IRetryPolicy, remove from here.
+        public final static Quadruple<Boolean, Boolean, Duration, Integer> INITIAL_ARGUMENT_VALUE_POLICY_ARG = Quadruple.with(false, false,
+            Duration.ofSeconds(60), 0);
+
+        public GoneRetryPolicy(
+            RxDocumentServiceRequest request,
+            Integer waitTimeInSeconds,
+            StopWatch durationTimer) {
+
+            checkNotNull(request, "request must not be null.");
+            this.request = request;
+            this.durationTimer = durationTimer;
+            this.waitTimeInSeconds = waitTimeInSeconds != null ? waitTimeInSeconds : DEFAULT_WAIT_TIME_IN_SECONDS;
+        }
+
+        private boolean isRetryableException(Exception exception) {
+            if (exception instanceof GoneException ||
+                exception instanceof RetryWithException ||
+                exception instanceof PartitionIsMigratingException ||
+                exception instanceof PartitionKeyRangeIsSplittingException) {
+
+                return true;
+            }
+
+            if (exception instanceof InvalidPartitionException) {
+                return this.request.getPartitionKeyRangeIdentity() == null ||
+                    this.request.getPartitionKeyRangeIdentity().getCollectionRid() == null;
+            }
+
+            return false;
+        }
+
+        private CosmosException logAndWrapExceptionWithLastRetryWithException(Exception exception) {
+            RetryWithException lastRetryWithException = this.request.getLastRetryWithException();
+
+            String exceptionType;
+            if (exception instanceof GoneException) {
+                exceptionType = "GoneException";
+            } else if (exception instanceof PartitionKeyRangeGoneException) {
+                exceptionType = "PartitionKeyRangeGoneException";
+            } else if (exception instanceof  InvalidPartitionException) {
+                exceptionType = "InvalidPartitionException";
+            } else if (exception instanceof  PartitionKeyRangeIsSplittingException) {
+                exceptionType = "PartitionKeyRangeIsSplittingException";
+            } else if (exception instanceof CosmosException) {
+                logger.warn("Received CosmosException after backoff/retry. Will fail the request.",
+                    exception);
+
+                return (CosmosException)exception;
+            } else {
+                throw new IllegalStateException("Invalid exception type", exception);
+            }
+
+            if (lastRetryWithException != null) {
```

Comment:
why do we still need to handle retryWithException here? don't we have a standalone retryPolicy for that?

Reply:
Problem is that the lastRetryWithException is state needed by both RetryWithRetryPolicy and GoneRetryPolicy - but you are right pushing it to RxDocumentServiceRequest was convenient but a hack. Changed it in latest iteration.